### PR TITLE
New remote window

### PR DESCRIPTION
**Pull request by chrisnojima**
This is a new implementation of how we work with remote windows

## Old Way
The app has a RemoteManager (https://github.com/keybase/client/blob/master/shared/desktop/renderer/remote-manager.js#L8) which is always rendered, but invisible. Remote manager houses RemoteTracker, RemotePinentry, RemoteUnlockFolders, etc. All proxies for a remote browser window.
Remote* were implemented using RemoteComponent (https://github.com/keybase/client/blob/master/shared/desktop/renderer/remote-component.desktop.js#L27).
RemoteComponent does
- Manages the BrowserWindow (creates, makes menus, loads the url)
- Tells the node process the window exists and needs props
- Renders an id to the app to help debugging

The way the data moves is, the RemoteComponent has a RemoteStore (https://github.com/keybase/client/blob/master/shared/desktop/renderer/remote-store.desktop.js#L4)
It asks the main node process for store updates
It runs thunks in its own process (hence us needing to run a full engine in the browser windows)
It sends plain actions to the main node process to send to the main window to process

The way we send data to the RemoteStores is:
The main desktop renderer (https://github.com/keybase/client/blob/master/shared/desktop/renderer/index.js#L111)
subscribes to the store and (in a throttled way) calculates the subset of the store the remote components cares about (each one exposes a 'selector') (example here: https://github.com/keybase/client/blob/master/shared/unlock-folders/selector.js)
It keeps a JSON.stringify() version to see if that has changed, if it has it sends the entire store to all listening components

Pros:
- As far as the components are concerned they get the actual store and just render themselves as if they're in the main app
- Can use thunks
- Can sometimes use HiddenString (in thunks but not as plain actions)

Cons:
- Can't send immutable over the wire so we can't use immutable stores for any remote components
- Windows send everything into the main app which then sends it out
- Have to keep selectors in sync with the needs of the components
- Some actions are handled in the main window, some in the remote window

## New Way

The app has a RemoteManager which is always rendered, but invisible. Remote manager houses RemoteTracker, RemotePinentry, RemoteUnlockFolders, etc. All proxies for a remote browser window. The key difference is we actually render props to these components and those props will ultimately be sent across the wire. This means its more WYSIWYG. We can render nothing (like in prod), we can render the props with JSON.stringify() , we can actually render the component inside of main. We can leverage connect/reselect/whatever to only send updates when the props actually change.

Remote* are implemented as a series of HOCs. As an example
RemotePinentry (https://github.com/keybase/client/pull/9695/files#diff-9f8790600c92751dd4481bab2e373aa0R37)
```
const RemotePinentry = compose(connect(pinentryMapStateToProps, () => ({})), RemoteWindow, RemoteConnector)(
 PrintDebug
 )
```
Which 
1. Does a mapStateToProps to get the actual props to send over the wire
1. We don't ever have dispatch here (cause all these props get sent over the wire)
1. Uses RemoteWindow (https://github.com/keybase/client/pull/9695/files#diff-f19fa7f3ac368e1b1d12141c74e86366R48). RemoteWindow only handles keeping the BrowserWindow live while this proxy is mounted
1. Uses RemoteConnector (https://github.com/keybase/client/pull/9695/files#diff-c9152cd98ddc0f55de765a8fd70b67bdR17) handles sending props changes over the wire (directly to the main window)

So the way data moves now is your RemotePinentry is like a regular component in the main app, it has a connect(), but instead of drawing itself on the screen, on render it sends its props over the wire to its BrowserWindow

The RemotePinentry has a RemoteStore(2) which sends all actions to the main window.
The components have a remote-container which is aware that its store is not a regular store but is a remote store (https://github.com/keybase/client/pull/9695/files#diff-e5f63b95c6badc5a78b7c50aeafffbf0R1)
This thing
1. Makes the mapStateToDispatch actions so the remote component can send plain actions over the wire
1. Is a plain pass through for the props (not strictly necessary but likely what we want)

so tldr;
A RemotePinentry renders its plain props through some HOCs to make a BrowserWindow and send those props to that BrowserWindow
The pinentry component has a special connector that takes those props it got over the wire and just sends it to the dumb component and dispatches plain actions

Pros:
- Remotes directly communicate with the main window
- Actions all happen in the main window
- Immutable stores everywhere
- Better separation of concerns
Cons:
- Can only send plain action objects, basically requiring sagas
- Can't send complex objects in actions (no hiddenstring)
- Avatar can't make rpc calls, but really thats very strange so its more reduxy now

### CODE REVIEW

File: shared/actions/pinentry.js [hunk 1]
```diff
@@ -1,76 +1,96 @@
 // @flow
 import * as PinentryGen from '../actions/pinentry-gen'
```
Comment (chrisnojima) [inline shared/actions/pinentry.js]: saga-ize and use entity helpers

File: shared/desktop/renderer/index.js [hunk 2]
```diff
@@ -65,6 +63,10 @@ function setupApp(store) {
 
   setupContextMenu(electron.remote.getCurrentWindow())
 
+  // Tell the main window some remote window needs its props
```
Comment (MarcoPolo) [inline shared/desktop/renderer/index.js]: This comment is wrong? this listens for this event

File: shared/desktop/renderer/remote-connector.desktop.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+// export default RemoteComponent
+export default function RemoteConnector(ComposedComponent: any) {
+  class RemoteConnected extends React.PureComponent<Props> {
+    _sendProps = () => {
+      if (this.props.remoteWindow) {
+        try {
+          const props = this._propsToSend()
+          console.log('aaa RemoteConnector sending props', JSON.stringify(props, null, 2))
```
Comment (MarcoPolo) [inline shared/desktop/renderer/remote-connector.desktop.js]: reword?

File: shared/desktop/renderer/remote-connector.desktop.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    componentDidMount() {
+      ipcRenderer.on('remoteWindowWantsProps', this._onNeedProps)
+    }
+    componentWillUnmount() {
+      ipcRenderer.removeListener('remoteWindowWantsProps', this._onNeedProps)
+    }
+
+    render() {
+      this._sendProps()
```
Comment (MarcoPolo) [inline shared/desktop/renderer/remote-connector.desktop.js]: Should we be doing this sideeffect in the render?

**Comment by keybase-ci-visdiff** [comment]
The commits 95691afd0d36...76d0cc861475 introduce visual changes on linux. <details><summary>:mag_right: 1 removed, 2 changed</summary>

 * Changed: **Pinentry-Passphrase_With_Error** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1511821159478-95691afd0d36-76d0cc861475-linux/Pinentry-Passphrase_With_Error-95691afd0d36.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1511821159478-95691afd0d36-76d0cc861475-linux/Pinentry-Passphrase_With_Error-76d0cc861475.png)  
   Pinentry-Passphrase_With_Error comparison
 * Changed: **Pinentry-Passphrase_Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1511821159478-95691afd0d36-76d0cc861475-linux/Pinentry-Passphrase_Normal-95691afd0d36.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1511821159478-95691afd0d36-76d0cc861475-linux/Pinentry-Passphrase_Normal-76d0cc861475.png)  
   Pinentry-Passphrase_Normal comparison
 * Removed: **Pinentry-Passphrase_Save_in_keychain** [(view original)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1511821159478-95691afd0d36-76d0cc861475-linux/Pinentry-Passphrase_Save_in_keychain-95691afd0d36.png)
</summary>

**Comment by keybase-ci-visdiff** [comment]
The commits 95691afd0d36...133c975380f5 introduce visual changes on linux. <details><summary>:mag_right: 1 removed, 2 changed</summary>

 * Changed: **Pinentry-Passphrase_With_Error** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1511885098929-95691afd0d36-133c975380f5-linux/Pinentry-Passphrase_With_Error-95691afd0d36.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1511885098929-95691afd0d36-133c975380f5-linux/Pinentry-Passphrase_With_Error-133c975380f5.png)  
   Pinentry-Passphrase_With_Error comparison
 * Changed: **Pinentry-Passphrase_Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1511885098929-95691afd0d36-133c975380f5-linux/Pinentry-Passphrase_Normal-95691afd0d36.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1511885098929-95691afd0d36-133c975380f5-linux/Pinentry-Passphrase_Normal-133c975380f5.png)  
   Pinentry-Passphrase_Normal comparison
 * Removed: **Pinentry-Passphrase_Save_in_keychain** [(view original)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1511885098929-95691afd0d36-133c975380f5-linux/Pinentry-Passphrase_Save_in_keychain-95691afd0d36.png)
</summary>

File: shared/desktop/renderer/remote-connector.desktop.js (hunk 4 above)
Comment (chrisnojima) [reply shared/desktop/renderer/remote-connector.desktop.js]: yes, this isn't a real component, it just sends data over the wire and thats it

**Comment by chrisnojima** [comment]
@MarcoPolo updated the description, PTAL and let me know if this makes sense

**Comment by chrisnojima** [comment]
- [ ] remove selector params
- [ ] remove printdebug
- [ ] remove old connector things
- [ ] node needs engine?

**Comment by keybase-ci-visdiff** [comment]
The commits 95691afd0d36...457b3fc5653c introduce visual changes on linux. <details><summary>:mag_right: 1 removed, 2 changed</summary>

 * Changed: **Pinentry-Passphrase_With_Error** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1511925223129-95691afd0d36-457b3fc5653c-linux/Pinentry-Passphrase_With_Error-95691afd0d36.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1511925223129-95691afd0d36-457b3fc5653c-linux/Pinentry-Passphrase_With_Error-457b3fc5653c.png)  
   Pinentry-Passphrase_With_Error comparison
 * Changed: **Pinentry-Passphrase_Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1511925223129-95691afd0d36-457b3fc5653c-linux/Pinentry-Passphrase_Normal-95691afd0d36.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1511925223129-95691afd0d36-457b3fc5653c-linux/Pinentry-Passphrase_Normal-457b3fc5653c.png)  
   Pinentry-Passphrase_Normal comparison
 * Removed: **Pinentry-Passphrase_Save_in_keychain** [(view original)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1511925223129-95691afd0d36-457b3fc5653c-linux/Pinentry-Passphrase_Save_in_keychain-95691afd0d36.png)
</summary>

**Comment by MarcoPolo** [comment]
Description makes sense, going to start reviewing

**Comment by chrisnojima** [comment]
Hold up cause I'm still working through it

File: shared/actions/chat/index.js [hunk 5]
```diff
@@ -28,66 +28,66 @@ function* _incomingTyping(action: ChatGen.IncomingTypingPayload): Saga.SagaGener
 }
 
 function* _setupChatHandlers(): Saga.SagaGenerator<any, any> {
-  engine().setIncomingActionCreator('chat.1.NotifyChat.NewChatActivity', ({activity}) =>
-    ChatGen.createIncomingMessage({activity})
-  )
+  engine().setIncomingActionCreators('chat.1.NotifyChat.NewChatActivity', ({activity}) => [
```
Comment (chrisnojima) [inline shared/actions/chat/index.js]: use action creators, also it takes an array now vs a single action

File: shared/actions/config.js [hunk 6]
```diff
@@ -196,12 +198,97 @@ function _bootstrapSuccess(action: ConfigGen.BootstrapSuccessPayload, state: Typ
   return Saga.all(actions)
 }
 
+function _pgpSecurityModelChangeMessageSaga() {
```
Comment (chrisnojima) [inline shared/actions/config.js]: removed the pgp store since it literally stores one boolean

File: shared/actions/config.js [hunk 7]
```diff
@@ -196,12 +198,97 @@ function _bootstrapSuccess(action: ConfigGen.BootstrapSuccessPayload, state: Typ
...
+    }
+    map[name] = urlMap
+    return map
+  }, {})
+
+  return Saga.put(ConfigGen.createLoadedAvatars({nameToUrlMap}))
+}
+
+let _avatarsToLoad = {}
+function* _loadAvatars(action: ConfigGen.LoadAvatarsPayload) {
```
Comment (chrisnojima) [inline shared/actions/config.js]: avatars now use redux instead of their own little weird thing (util/pictures and registering callbacks back and forth and etc)

File: shared/actions/config.js [hunk 8]
```diff
@@ -196,12 +198,97 @@ function _bootstrapSuccess(action: ConfigGen.BootstrapSuccessPayload, state: Typ
...
+  if (names.length) {
+    yield Saga.put({
+      payload: {endpoint: 'image/team_avatar_lookups', key: 'team_names', names},
+      type: '_loadAvatarHelper',
+    })
+  }
+}
+
+// Every minute we clear out any avatars that might have errored out
+function* _periodicAvatarCacheClear(): Generator<any, void, any> {
```
Comment (chrisnojima) [inline shared/actions/config.js]: we now periodically go through and cleanup missing avatars incase those errors could resolve themselves

File: shared/actions/tracker.js [hunk 9]
```diff
@@ -12,45 +12,21 @@ import {isMobile} from '../constants/platform'
...
-  const setupUserChangedHandlerHelper = (dispatch: Dispatch, getState: () => TypedState) => {
-    engine().setIncomingHandler('keybase.1.NotifyUsers.userChanged', ({uid}) => {
-      dispatch(TrackerGen.createCacheIdentify({uid, goodTill: 0}))
-      const username = _getUsername(uid, getState())
-      if (username) {
-        dispatch(getProfile(username))
-      }
-    })
+// Send a heartbeat while trackers are still open
+function* _trackerTimer(): Generator<any, void, any> {
```
Comment (chrisnojima) [inline shared/actions/tracker.js]: instead of explicitly trying to bookkeep this and keeping track of timers we just spawn a saga which periodically checks and sends the rpc if there are any trackers alive

File: shared/actions/tracker.js [hunk 10]
```diff
@@ -621,37 +531,118 @@ const _fillFolders = (username: string) => (dispatch: Dispatch, getState: () =>
...
+  const {uid} = action.payload
+  const actions = [TrackerGen.createCacheIdentify({goodTill: 0, uid})]
+  const username = _getUsername(uid, state)
+  if (username) {
+    actions.push(TrackerGen.createGetProfile({username}))
+  }
+  return actions
 }
+
+function _setupTrackerHandlers() {
```
Comment (chrisnojima) [inline shared/actions/tracker.js]: we had 2 functions for this but now its just one

File: shared/actions/tracker.js [hunk 11]
```diff
@@ -621,37 +531,118 @@ const _fillFolders = (username: string) => (dispatch: Dispatch, getState: () =>
...
+      .then(response => {
+        console.log('Registered identify ui')
+      })
+      .catch(error => {
+        console.warn('error in registering identify ui: ', error)
+      })
+  })
+
+  // TODO get rid of getState here
+  engine().setIncomingActionCreators(
```
Comment (chrisnojima) [inline shared/actions/tracker.js]: i'm plumbing through dispatch and getState to this action creator temporarily. This is because this inner thing is using an incomingCallMap and not a nicer saga thing. this is the singular use of this

File: shared/app/main-shared.native.js [hunk 12]
```diff
@@ -8,8 +8,6 @@ import loadPerf from '../util/load-perf'
 import hello from '../util/hello'
 import {connect, type TypedState} from '../util/container'
 import debounce from 'lodash/debounce'
-import {getUserImageMap, loadUserImageMap, getTeamImageMap, loadTeamImageMap} from '../util/pictures'
```
Comment (chrisnojima) [inline shared/app/main-shared.native.js]: this was this nasty avatar hookup code that was required

File: shared/common-adapters/avatar.js [hunk 13]
```diff
@@ -125,200 +121,115 @@ const followSizeToStyle = {
...
-const initLoad = (loadAvatar: AvatarLoad, loadTeam: TeamAvatarLoad) => {
-  _loadAvatarToURL = loadAvatar
-  _loadTeamAvatarToURL = loadTeam
-}
-
-export default Avatar
-export {initLoad, initLookup}
+// We could theoretically bundle all these requests into one action but this is simpler for now
 export type {AvatarSize}
+export default compose(
```
Comment (chrisnojima) [inline shared/common-adapters/avatar.js]: This had a lot of complexity which is gone now 🔥

File: shared/common-adapters/avatar.render.desktop.js [hunk 14]
```diff
@@ -56,19 +56,13 @@ const BackgroundDiv = glamorous.div(
     top: backgroundOffset,
   },
   props => ({
-    backgroundColor: props.loaded ? globalColors.white : props.loadingColor || globalColors.lightGrey,
```
Comment (chrisnojima) [inline shared/common-adapters/avatar.render.desktop.js]: turns out the loading flow got broken awhile ago. what happens is it starts out with the placeholder avatar, it tries to load it and it sees its loaded and so loaded is basically always true. so we can rip this all out and it looks 100% the same

File: shared/common-adapters/user-bio.js.flow [hunk 15]
```diff
@@ -23,8 +23,6 @@ export type Props = {
   editFns?: ?BioEditFns,
   onAvatarLoaded?: () => void,
   onClickAvatar: (username: ?string) => void,
-  onClickFollowers: (username: ?string) => void,
```
Comment (chrisnojima) [inline shared/common-adapters/user-bio.js.flow]: this was actually never hooked up to anything so i removed it

File: shared/desktop/app/menu-bar.js [hunk 16]
```diff
@@ -69,61 +67,15 @@ export default function() {
...
   mb.on('ready', () => {
+    menubarWindowIDCallback(mb.window.id)
     // Hack: open widget when left/right/double clicked
     mb.tray.on('right-click', (e, bounds) => {
       e.preventDefault()
       setImmediate(() => mb.tray.emit('click', {...e}, {...bounds}))
     })
     mb.tray.on('double-click', e => e.preventDefault())
 
-    const webContents = mb.window.webContents
```
Comment (chrisnojima) [inline shared/desktop/app/menu-bar.js]: this is basically copy past of remote windows in general so we can kill it 🔥

File: shared/desktop/renderer/index.js (hunk 2 above)
Comment (chrisnojima) [reply shared/desktop/renderer/index.js]: it listens for the event and sends it to the store, which then tells the main window

**Comment by chrisnojima** [comment]
- [ ] Merge after the next release

File: shared/common-adapters/avatar.render.desktop.js [hunk 17]
```diff
@@ -56,19 +56,13 @@ const BackgroundDiv = glamorous.div(
...
+class Background extends React.PureComponent<{borderRadius: any}> {
   render() {
-    return (
-      <BackgroundDiv
-        loaded={this.props.loaded}
-        loadingColor={this.props.loadingColor}
-        borderRadius={this.props.borderRadius}
-      />
-    )
+    return <BackgroundDiv borderRadius={this.props.borderRadius} />
```
Comment (buoyad) [inline shared/common-adapters/avatar.render.desktop.js]: Do you still want to pass `loadingColor` here?

File: shared/common-adapters/avatar.render.desktop.js [hunk 18]
```diff
@@ -209,12 +138,7 @@ class AvatarRender extends React.PureComponent<Props, State> {
           ...this.props.style,
         }}
       >
-        {!this.props.skipBackground &&
-          <Background
-            loaded={this.state.loaded}
-            loadingColor={this.props.loadingColor}
-            borderRadius={borderRadius}
-          />}
+        {!this.props.skipBackground && <Background borderRadius={borderRadius} />}
```
Comment (buoyad) [inline shared/common-adapters/avatar.render.desktop.js]: here too

File: shared/desktop/remote/sync-browser-window.desktop.js [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+        async: false,
+        src: hotPath('component-loader.bundle.js'),
+      },
+    ],
+    selectorParams,
+    title,
+  })
+}
+
+function BrowserWindowSync(ComposedComponent: any) {
```
Comment (MarcoPolo) [inline shared/desktop/remote/sync-browser-window.desktop.js]: I think it would be helpful to easily understand interdependencies if this was typed:

Type is probably something like:

```
function BrowserWindowSync<P>(ComposeComponent: React.Component<{...P, remteWindow: BrowserWindow}>) {
...
```

Comment (MarcoPolo) [reply shared/desktop/remote/sync-browser-window.desktop.js]: 🐑  🐑  Also it might be nice to just pass down what we need right now, so it's a bit clearer what consumers of the prop can do. i.e. just pass down `remoteWindowEmit: (eventName, eventArgs) => void`

File: shared/desktop/remote/sync-props.desktop.js [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+const ipcRenderer = electron.ipcRenderer
+const BrowserWindow = remote.BrowserWindow
+
+type Props = {
+  selectorParams: ?string,
+  component: string,
+  remoteWindow: ?BrowserWindow,
+}
+
+function SyncProps(ComposedComponent: any) {
```
Comment (MarcoPolo) [inline shared/desktop/remote/sync-props.desktop.js]: With the same reasoning as typing BrowserWindowSync, this one would be:

```
function SyncProps<P>(ComposedComponent: <P & {selectorParams: ?string, component: string}>) {
  class RemoteConnected extends React.PureComponent<Props> {
...
```

File: shared/actions/pinentry.js [hunk 21]
```diff
@@ -1,76 +1,84 @@
...
+  )
 }
 
-function uglyResponse(sessionID: number, result: any, err: ?any): void {
-  const response = uglySessionIDResponseMapper[sessionID]
+function _respond(sessionID: number, result: any, err: ?any): void {
+  const sessionKey = String(sessionID)
+  const response = sessionIDToResponse[sessionKey]
   if (response == null) {
     console.log('lost response reference')
```
Comment (MarcoPolo) [inline shared/actions/pinentry.js]: When the new logger stuff gets merged this should probably be logger.error becuase this should never happen

File: shared/menubar/index.desktop.js [hunk 22]
```diff
@@ -256,15 +270,12 @@ const BadgeIcon = ({
...
-MenubarRender.defaultProps = {
-  openToPrivate: true,
-  openWithMenuShowing: false,
-}
-
 const borderRadius = 4
 
 const stylesContainer = {
   ...globalStyles.flexBoxColumn,
+  width: '100%',
```
Comment (buoyad) [inline shared/menubar/index.desktop.js]: This component is getting an extra spacing on the top now, maybe from this?

image

File: shared/desktop/remote/component-loader.js [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+  _onGotProps = () => {
+    // Show when we get props, unless its the menubar
+    if (this._window && this.props.component !== 'menubar') {
+      this._window.show()
+    }
+  }
+
+  _getComponent = (key: string) => {
+    switch (key) {
+      case 'puregeMessage':
```
Comment (buoyad) [inline shared/desktop/remote/component-loader.js]: `puregeMessage` -> `purgeMessage`

File: shared/desktop/remote/sync-browser-window.desktop.js [hunk 24]
```diff
@@ -0,0 +1,162 @@
...
+      // Don't forward our internal props
+      const {windowOpts, positionBottomRight, title, ...props} = this.props
+      return <ComposedComponent {...props} remoteWindow={this.state.remoteWindow} />
+    }
+  }
+
+  return RemoteWindowComponent
+}
+
+export default BrowserWindowSync
```
Comment (buoyad) [inline shared/desktop/remote/sync-browser-window.desktop.js]: This is called `SyncBrowserWindow` everywhere it's imported

File: shared/engine/index.js.flow [hunk 25]
```diff
@@ -19,7 +20,10 @@ declare export class Engine {
   listenOnDisconnect(key: string, f: () => void): void,
   hasEverConnected(): void,
   setIncomingHandler(name: string, callback: Function): void,
-  setIncomingActionCreator(method: string, actionCreator: (param: Object, response: ?Object) => ?any): void,
+  setIncomingActionCreators(
+    method: string,
+    actionCreator: (param: Object, response: ?Object, dispatch: Dispatch, getState: () => TypedState) => ?any
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/engine/index.js.flow]: yes its used in one place which is why i added it. really i have to fixup some incomingcallmap stuff then we could get rid of it

File: shared/common-adapters/avatar.render.desktop.js (hunk 17 above)
Comment (chrisnojima) [reply shared/common-adapters/avatar.render.desktop.js]: i don't think this is ever set

File: shared/actions/config.js [hunk 26]
```diff
@@ -196,12 +198,97 @@ function _bootstrapSuccess(action: ConfigGen.BootstrapSuccessPayload, state: Typ
...
+}
+
+function _loadAvatarHelper(action: {payload: {names: Array<string>, endpoint: string, key: string}}) {
+  const {names, endpoint, key} = action.payload
+  return Saga.all([
+    Saga.call(RPCTypes.apiserverGetRpcPromise, {
+      args: [{key, value: names.join(',')}, {key: 'formats', value: 'square_360,square_200,square_40'}],
+      endpoint,
+    }),
+    Promise.resolve(names),
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/actions/config.js]: its not a passthrough, its one of the params and the output

File: shared/desktop/remote/sync-browser-window.desktop.js (hunk 19 above)
Comment (chrisnojima) [reply shared/desktop/remote/sync-browser-window.desktop.js]: but this hoc doesn't understand what its children needs. Yes, thats all the syncProps might need but this isn't true in general. this hoc is all about managing a BrowserWindow so thats what it passes down

File: shared/desktop/remote/sync-browser-window.desktop.js [hunk 27]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    componentWillUnmount() {
+      this._mounted = false
+      this._closeBrowserWindow()
+    }
+
+    render() {
+      // Don't forward our internal props
+      const {windowOpts, positionBottomRight, title, ...props} = this.props
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/desktop/remote/sync-browser-window.desktop.js]: renamed them all, including selectorParams cause that was an old holdover

File: shared/actions/config.js [hunk 28]
```diff
@@ -199,12 +201,101 @@ function _bootstrapSuccess(action: ConfigGen.BootstrapSuccessPayload, state: Typ
...
+    }
+    map[name] = urlMap
+    return map
+  }, {})
+
+  return Saga.put(ConfigGen.createLoadedAvatars({nameToUrlMap}))
+}
+
+function _validUsernames(names: Array<string>) {
+  return names.filter(name => !!name.match(/^([.a-z0-9_-]{1,1000})$/i))
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/actions/config.js]: i just wanted to keep the existing behavior

File: shared/actions/unlock-folders.js [hunk 29]
```diff
@@ -37,16 +38,35 @@ const registerRekeyListener = () => (dispatch: Dispatch) => {
...
+          problemSet: problemSetDevices.problemSet,
+          sessionID,
+        }),
+      ]
+    }
   )
 
   // else we get this also as part of delegateRekeyUI
-  engine().setIncomingHandler('keybase.1.rekeyUI.delegateRekeyUI', (param: any, response: ?Object) => {
+  engine().setIncomingActionCreators('keybase.1.rekeyUI.delegateRekeyUI', (_, response, _dispatch) => {
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/actions/unlock-folders.js]: originally cause this is this weird one thats passed in from engine but, i'll just remove it

File: shared/actions/config.js (hunk 28 above)
[the remark this answers is not in the document]
Comment (buoyad) [reply shared/actions/config.js]: ah yeah, I wasn't sure if this was existing logic

File: shared/menubar/index.desktop.js (hunk 22 above)
Comment (chrisnojima) [reply shared/menubar/index.desktop.js]: good catch! turns out this was a teeny bit more complicated. see the last commit

**Comment by chrisnojima** [comment]
@buoyad @MarcoPolo updated ptal

**Comment by chrisnojima** [comment]
ok i've merged this down. i'll merge it monday morning
